server/verify.go

```diff
@@ -48,27 +51,45 @@ func VerifyAttestation(attestation *pb.Attestation, opts VerifyOpts) (*pb.Machin
 		return nil, err
 	}
 
-	// For now, we only verify SHA256 based Quotes, PCRs, and Events
-	var quote *tpmpb.Quote
-	for _, q := range attestation.GetQuotes() {
-		if q.GetPcrs().GetHash() != tpmpb.HashAlgo_SHA256 {
-			continue
-		}
-		quote = q
+	// Verify the signing hash algorithm
+	signHashAlg, err := internal.GetSigningHashAlg(akPubArea)
+	if err != nil {
+		return nil, fmt.Errorf("bad AK public area: %w", err)
 	}
-	if quote == nil {
-		return nil, fmt.Errorf("attestation does not contain a SHA256 quote")
+	if err = checkHashAlgSupported(signHashAlg, opts); err != nil {
+		return nil, fmt.Errorf("in AK public area: %w", err)
 	}
-	// Verify the Quote
-	if err = internal.VerifyQuote(quote, akPubKey, opts.Nonce); err != nil {
-		return nil, fmt.Errorf("failed to verify quote: %w", err)
+
+	// Attempt to replay the log against our PCRs in order of hash preference
+	var lastErr error
+	for _, quote := range supportedQuotes(attestation.GetQuotes()) {
+		// Verify the Quote
+		if err = internal.VerifyQuote(quote, akPubKey, opts.Nonce); err != nil {
+			lastErr = fmt.Errorf("failed to verify quote: %w", err)
+			continue
+		}
+
+		// Parse the event log and replay the events against the provided PCRs
+		pcrs := quote.GetPcrs()
+		state, err := ParseAndReplayEventLog(attestation.GetEventLog(), pcrs)
+		if err != nil {
+			lastErr = fmt.Errorf("failed to validate the event log: %w", err)
+			continue
+		}
+
+		// Verify the PCR hash algorithm
+		pcrHashAlg := tpm2.Algorithm(pcrs.GetHash())
```

Comment:
this should be towards the top before doing any of the verification/replay

Reply:
We have this check here (instead of at
the start of the loop) so that the user gets a "SHA-1 not supported"
error only if allowing SHA-1 support would actually allow the log
to be verified. This makes debugging failed verifications easier.

Added a comment explaining why.